[Tables] Tables batch

joheredi opened the pull request:
## API Level changes
* Adds a `createBatch` method which returns a `TablesBatch` object which is used to hold the operations to be performed in the batch.

* The `TablesBatch` object exposes the following methods:
   * createEntities - Adds an insert operation to the Batch for each of the entities in the array. Does not perform a service call
   * createEntity - Adds a single insert operation to the Batch. Does not perform a service call
   * deleteEntity - Adds a delete operation to the Batch. Does not perform a service call
   * updateEntity - Adds an update operation to the batch. Does not perform a service call
   * submitBatch - Submits the operations in the batch. **Does Perform a service call**

## Implementation details

### Crafting the Batch request
- Swagger doesn't define the batch operation so it needs to be hand-crafted
- Batch request sends all the operations in the body, each operation is bounded by `changeset_<GUID>` and the whole Batch is bounded by `batch_<GUID>`.
   - Each operation is described with the HTTP request
   - When building the batch, I create a new internal client with a Pipeline that intercepts the operations and instead of sending them to the service it appends them to the current batch body

**Sample Batch**
```
POST https://myaccount.table.core.windows.net/$batch HTTP/1.1  
x-ms-version: 2013-08-15  
Accept-Charset: UTF-8  
DataServiceVersion: 3.0;  
MaxDataServiceVersion: 3.0;NetFx  
Content-Type: multipart/mixed; boundary=batch_a1e9d677-b28b-435e-a89e-87e6a768a431  
x-ms-date: Mon, 14 Oct 2013 18:25:49 GMT  
Authorization: SharedKey myaccount:50daR38MtfezvbMdKrGJVN+8sjDSn+AaA=  
Host: 127.0.0.1:10002  
Content-Length: 1323  
Connection: Keep-Alive  
  
--batch_a1e9d677-b28b-435e-a89e-87e6a768a431  
Content-Type: multipart/mixed; boundary=changeset_8a28b620-b4bb-458c-a177-0959fb14c977  
  
--changeset_8a28b620-b4bb-458c-a177-0959fb14c977  
Content-Type: application/http  
Content-Transfer-Encoding: binary  
  
POST https://myaccount.table.core.windows.net/Blogs HTTP/1.1  
Content-Type: application/json  
Accept: application/json;odata=minimalmetadata  
Prefer: return-no-content  
DataServiceVersion: 3.0;  
  
{"PartitionKey":"Channel_19", "RowKey":"1", "Rating":9, "Text":".NET..."}  
--changeset_8a28b620-b4bb-458c-a177-0959fb14c977  
Content-Type: application/http  
Content-Transfer-Encoding: binary  
  
POST https://myaccount.table.core.windows.net/Blogs HTTP/1.1  
Content-Type: application/json  
Accept: application/json;odata=minimalmetadata  
Prefer: return-no-content  
DataServiceVersion: 3.0;  
  
{"PartitionKey":"Channel_17", "RowKey":"2", "Rating":9, "Text":"Azure..."}  
--changeset_8a28b620-b4bb-458c-a177-0959fb14c977  
Content-Type: application/http  
Content-Transfer-Encoding: binary  
  
MERGE https://myaccount.table.core.windows.net/Blogs(PartitionKey='Channel_17', RowKey='3') HTTP/1.1  
Content-Type: application/json  
Accept: application/json;odata=minimalmetadata  
DataServiceVersion: 3.0;  
  
{"PartitionKey":"Channel_19", "RowKey":"3", "Rating":9, "Text":"PDC 2008..."}  
  
--changeset_8a28b620-b4bb-458c-a177-0959fb14c977--  
--batch_a1e9d677-b28b-435e-a89e-87e6a768a431  
```

joheredi commented on sdk/tables/data-tables/review/data-tables.api.md:
```diff
@@ -499,6 +524,7 @@ export class TableServiceClient {
 export type TableServiceClientOptions = PipelineOptions & {
     endpoint?: string;
     version?: string;
+    requestPolicyFactories?: RequestPolicyFactory[] | ((defaultRequestPolicyFactories: RequestPolicyFactory[]) => void | RequestPolicyFactory[]);
```
Added this in order to be able to create a client with a pipeline that allows intercepting requests to build the batch

xirzec replied:
Why not have `TableServiceClientOptions` take an `InnerBatchRequest` instead? That way you can have `TableClient` assemble the right policies instead of having them be passed in by the batch itself.

joheredi replied:
I'm trying this out, the only thing I'm not sure about is if we want to expose InnerBatchRequest. What do you think?

joheredi commented on sdk/tables/data-tables/src/TableClient.ts:
```diff
@@ -38,13 +41,17 @@ import { createPipelineFromOptions, InternalPipelineOptions } from "@azure/core-
 export class TableClient {
   private table: Table;
+  private url: string;
+  private credential: TablesSharedKeyCredentialLike | undefined;
```
Since the batch request is hand crafted, we need access to these 2 properties when sending the batch request later on

joheredi commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+  return `${urlParsed.getScheme()}://${urlParsed.getHost()}${urlParsed.getPath()}`;
+}
+
+class BatchHeaderFilterPolicy extends BaseRequestPolicy {
```
These classes would be moved to their own files

joheredi commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+  }
+}
+
+class BatchRequestAssemblePolicy extends BaseRequestPolicy {
```
This is the policy that intercepts the request and appends it to the batch body instead of sending it

joheredi commented on sdk/tables/data-tables/src/TablesSharedKeyCredentialPolicy.ts:
```diff
@@ -60,70 +60,66 @@ export class TablesSharedKeyCredentialPolicy extends BaseRequestPolicy {
    * @returns {WebResource}
    */
-  protected signRequest(request: WebResourceLike): WebResource {
+  public signRequest(request: WebResourceLike): WebResource {
```
I had to refactor these methods a bit to be able to reuse `getAuthorizationHeader` when hand-crafting the batch request

joheredi commented on sdk/tables/data-tables/test/integration/tableclient.spec.ts:
```diff
@@ -312,4 +312,30 @@ describe("TableClient", () => {
+      await recorder.stop();
+    });
+
+    it.only("should send a set of create batch operations", async () => {
```
will remove only

xirzec commented on sdk/tables/data-tables/samples/typescript/src/batchOperations.ts:
```diff
@@ -0,0 +1,77 @@
+
+  console.log(response);
+
+  // await client.delete();
```
is this not needed?

ellismg replied:
Similarly, the commented out create call on L21 is also not needed (as the comment above says it expects the table to exist).  I suspect this was here to make it easier to run the sample over and over during development and get back to a clean state each time.

xirzec commented on sdk/tables/data-tables/samples/typescript/src/batchOperations.ts:
```diff
@@ -0,0 +1,77 @@
+  // await client.delete();
+}
+
+interface Entity {
```
I'd put this up near the top of the file

xirzec commented on sdk/tables/data-tables/samples/typescript/src/batchOperations.ts:
```diff
@@ -0,0 +1,77 @@
+  quantity: number;
+}
+
+export async function main() {
+  await batchOperations();
+}
```
This doesn't do anything. You could just call `batchOperations()` directly in the below code instead of `main` and it'd do the same thing.

xirzec commented on sdk/tables/data-tables/src/TableClient.ts:
```diff
@@ -139,9 +147,14 @@ export class TableClient {
-    const pipeline = createPipelineFromOptions(internalPipelineOptions, credential);
+    let pipeline = createPipelineFromOptions(internalPipelineOptions, credential);
+
+    if (Array.isArray(clientOptions.requestPolicyFactories)) {
```
I don't get this piece of code, why have your constructor take a method that modifies the requestPolicyFactories if you're only going to do something with the array form? You could just declare `clientOptions.requestPolicyFactories` to be an array and be done with it.

Also, since all `createPipelineFromOptions` does is return this shape:

```ts
return {
    httpClient: pipelineOptions.httpClient,
    requestPolicyFactories
  };
```

You don't even need to call `createPipelineFromOptions` if you get passed `requestPolicyFactories`, it just does a bunch of wasted work.

joheredi replied:
Makes sense I'm taking your suggestion of getting taking InnerBatchRequest in TableServiceClientOptions. I'll check for that and conditionally call createPipelineFromOptions if it is not passed

xirzec commented on sdk/tables/data-tables/src/TableClient.ts:
```diff
@@ -473,6 +486,19 @@ export class TableClient {
+   * @param options The options parameters.
+   */
+  public createBatch(partitionKey: string): TablesBatch {
+    return createBatch(this.url, this.tableName, partitionKey, this.credential);
```
the naming gets a little confusing, having a class method named the same thing as your helper method.

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+import { getAuthorizationHeader } from "./TablesSharedKeyCredentialPolicy";
+import { HeaderConstants } from "./utils/constants";
+
+export interface TablesBatch {
```
why is it `TableClient` but `TablesBatch` (plural)?

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+  }
+
+  return {
+    partitionKey,
```
shouldn't this be constant/read-only? I'd be fine with you declaring it readonly on the interface rather than enforcing that in this object if you'd like. Although you could use Object.defineProperty to do it.

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+  return {
+    partitionKey,
+    async createEntity<T extends object>(entity: TableEntity<T>): Promise<void> {
+      if (entity.partitionKey !== this.partitionKey) {
```
seems like you could make a helper function to check this given an entity. You also don't need to use `this` here, since `partitionKey` is a scope variable.

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+        body,
+        disableJsonStringifyOnBody: true
+      };
+      console.log(body);
```
this definitely should get removed

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+      };
+      console.log(body);
+      const rawBatchResponse = await client.sendRequest(requestOptions);
+      console.log(rawBatchResponse.bodyAsText);
```
same here

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+        "x-ms-version": "2019-02-02",
+        "Accept-Charset": "UTF-8",
+        DataServiceVersion: "3.0;",
+        MaxDataServiceVersion: "3.0;NetFx",
```
isn't NetFx a .NET thing?

ellismg replied:
Suspect we need to do this because we are building the request ourselves (instead of allowing the generated client to do so).  Per https://docs.microsoft.com/en-us/rest/api/storageservices/setting-the-odata-data-service-version-headers it looks like the service expects.

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+    super(nextPolicy, options);
+  }
+  public async sendRequest(request: WebResourceLike): Promise<HttpOperationResponse> {
+    let xMsHeaderName = "";
+
+    for (const header of request.headers.headersArray()) {
+      if (header.name.toLowerCase() === HeaderConstants.X_MS_VERSION) {
+        xMsHeaderName = header.name;
+      }
+    }
+
+    if (xMsHeaderName !== "") {
+      request.headers.remove(xMsHeaderName); // The subrequests should not have the x-ms-version header.
+    }
```
You can use `contains` to check, but you really don't have to. Operations on `HttpHeaders` are case insensitive anyway.

```suggestion
    request.headers.remove(HeaderConstants.X_MS_VERSION); // The subrequests should not have the x-ms-version header.
```

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+    },
+    appendSubRequestToBody(request: WebResource) {
+      // Start to assemble sub request
+      this.body += [
```
I'd make the subrequest string and **then** append it to body, for readability. It might also be worthwhile to have the body be an array of strings until a final join before sending.

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+  const subRequestPrefix = `--${changesetBoundary}${HTTP_LINE_ENDING}${HeaderConstants.CONTENT_TYPE}: application/http${HTTP_LINE_ENDING}${HeaderConstants.CONTENT_TRANSFER_ENCODING}: binary`;
+  // Content-Type: multipart/mixed; boundary=changeset_{changesetGuid}
+  const multipartContentType = `multipart/mixed; boundary=changeset_${changesetGuid}`;
+  const chaangesetEnding = `--${changesetBoundary}--`;
```
typo: `changesetEnding`

xirzec commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+ * @param {string} url Source URL string
+ * @returns {(string | undefined)}
+ */
+export function getURLPathAndQuery(url: string): string | undefined {
```
I don't understand what this function does. It seems like it strips off the queryString, but the name doesn't make sense to me. Also what's the point in sanitizing the queryString if it's not returned or used?

xirzec reviewed (commented):
You have a really neat solution for assembling the batch request by overloading the pipeline, and I think we can hide some of this magic from customers so they don't get burned when we change up how pipelines work later.

I really like the overall feel of the API from a user perspective, most of my comments are on implementation.

ellismg commented on sdk/tables/data-tables/review/data-tables.api.md:
```diff
@@ -480,6 +489,22 @@ export interface TableResponseProperties {
+// @public (undocumented)
+export interface TablesBatch {
+    // (undocumented)
+    createEntities: <T extends object>(entitites: TableEntity<T>[]) => Promise<void>;
```
One possible solution here might be to have the batch itself just maintain a list of operations internally, so that all these CRUD methods just push the operations onto that.  Then, when submit batch is called, we "replay" these operations against the internal client to build the batch and then submit it.

Or perhaps, have these methods be void returning and just collect the promises from the internal call in an array and then the first thing SubmitBatch does is await them all in order?

xirzec replied:
Thank you @ellismg because I had this same thought early on but got distracted and forgot to leave a comment about it. It feels strange that they are async to me as well.

ellismg reviewed (commented):
Very cool!

Some quick comments on the API shape, I will dig into the implementation in more depth a little later today but wanted to get the ball rolling on feedback.

xirzec commented on sdk/tables/data-tables/review/data-tables.api.md:
```diff
@@ -304,11 +304,20 @@ export interface SignedIdentifier {
+// @public (undocumented)
+export interface TableBatchResponse {
+    // (undocumented)
+    getResponseForEntity: () => HttpResponse;
```
Yeah, this also seemed odd to me, like it's missing a parameter. (Again, the distractions!) Thanks for catching this.

joheredi commented on sdk/tables/data-tables/src/TablesBatch.ts:
```diff
@@ -0,0 +1,321 @@
+    body: `--${batchBoundary}${HTTP_LINE_ENDING}${HeaderConstants.CONTENT_TYPE}: multipart/mixed; boundary=changeset_${changesetGuid}${HTTP_LINE_ENDING}${HTTP_LINE_ENDING}`,
+    operationCount: 0,
+    createPipeline() {
+      const policyFactoryLength = 3;
```
Oh yeah you are right, this was horrible

joheredi commented:
@xirzec, @ellismg thanks a lot for your feedback :smile: I have addressed your comments and made this a real PR. When you get a chance can you please take another look?

joheredi commented on sdk/tables/data-tables/test/integration/tableclient.spec.ts:
```diff
@@ -347,4 +355,80 @@ describe("TableClient", () => {
     });
   });
+
+  if (isNode || isLiveMode()) {
```
There is a weird issue when executing these tests against browser and playback mode, live mode works fine, and test with any mode works for node.

I've filed https://github.com/Azure/azure-sdk-for-js/issues/11603 to investigate further. Meanwhile, I'm skipping the test if run in playback mode against the browser

joheredi commented:
/azp run js - tables - tests

joheredi commented:
/azp run js - tables - tests

joheredi commented:
/azp run js - tables - tests

ellismg commented on sdk/tables/data-tables/review/data-tables.api.md:
```diff
@@ -318,6 +350,20 @@ export class TableClient {
+    }
+
+// @public
+export interface TableClientLike {
```
Just wondering - do we need to export this as part of our public surface area?  It didn't seem to be referenced anywhere else in this review.md

xirzec replied:
Yeah I don't think we actually need to export this type.

joheredi replied:
You're right I just needed this to avoid a circular dependency in TableBatch, moving this to internalModels

ellismg commented on sdk/tables/data-tables/src/TableBatch.ts:
```diff
@@ -0,0 +1,294 @@
+   * Submits the operations in the batch
+   */
+  public async submitBatch(): Promise<any> {
+    await Promise.all(this.pendingOperations);
```
Is the order in which `Promise.all()` resolves the promises specified?  It was not obvious to me from reading MDN that this was the case.  Assuming that the resolution order is unspecified, should we instead just do an explicit loop here and await in order?  I'm slightly worried about the batch being built "out of order" and maybe that leading to something observable, although all the cases I can think of to observe this would involve things like "creating and deleting an entity as part of a batch".

xirzec replied:
Awaiting in order doesn't guarantee order of execution once they've been kicked off in parallel. It would only check if they were settled or not in order.

I think this should be safe, as you can't have conflicting operations within a single batch regardless of order, right?

joheredi replied:
Would it be safer to store an array of callbacks instead of promises, and then call/await each of them on submit? I wonder if we can run into race-conditions and get a malformed body

xirzec replied:
Hmm malformed in what way? No promise code could run simultaneously, so unless the code did something like 

```
// modify the body in a way that makes it incorrect
await something();
// modify the body in a way that makes it correct again, hope nobody else touched it
```

I think we should be okay?

xirzec replied:
Yeah it looks like the policy calls `appendSubRequestToBody` which synchronously adds something to the body array, and since we don't call `getHttpRequestBody` until all promises are settled (e.g. all body parts are added to the array), we'll have them all.

joheredi replied:
Since we are pushing to the array and not mutating the body, the worst that can happen is to have the operations added to the body in a different order as Matt mentioned.

I just checked and the service doesn't seem to allow multiple changes to the same rowKey within a batch so it looks like this probably wouldn't be an issue.

I think I'll keep the current approach as it has the extra benefit of computing start computing the body in the background as soon as the batch operations are called, so submitBatch will be faster than if we await each callback during submitBatch.

Does this make sense?


This is the error that the service gives us back

```javascript
{
  status: 202,
  subResponses: [
    {
      status: 400,
      body: {
        "odata.error": {
          code: "InvalidDuplicateRow",
          message: {
            lang: "en-US",
            value:
              "1:The batch request contains multiple changes with same row key. An entity can appear only once in a batch request. RequestId:456aeab3-b002-0014-0bef-986a9d000000 Time:2020-10-02T19:11:31.6622404Z",
          }
        }
      }
    }
  ]
}
```

ellismg replied:
Yeah, that makes sense to me.  Thanks for checking to see that the service rejects such cases.

ellismg commented on sdk/tables/data-tables/review/data-tables.api.md:
```diff
@@ -304,11 +304,43 @@ export interface SignedIdentifier {
+}
+
+// @public
+export interface TableBatchOptions {
```
Are there cases where folks would want to explicitly set these values, or is this simply because we need to do it for determinism in our tests with respect to the recordings?  If this is just for testing, I do wonder if there's some way for us to hide it, but to be honest, I don't know what that would look like.  Maybe Jeff has some idea.

xirzec replied:
We should be able to make the recorder smart enough to not require these if they are indeed just for testing.

joheredi replied:
I was able to get rid of these by stubbing generateUuid in the tests

ellismg reviewed (commented):
Great progress!  I had a few questions but overall this looks really good.  Really excited to see this land and delivered in the next preview!

xirzec commented on sdk/tables/data-tables/review/data-tables.api.md:
```diff
@@ -304,11 +304,43 @@ export interface SignedIdentifier {
+
+// @public
+export interface TableBatchResponse {
+    getResponseForEntity: (rowKey: string) => TableBatchEntityResponse | undefined;
```
If we do make it possible to type the entity:

```suggestion
    getResponseForEntity<T>: (rowKey: string) => TableBatchEntityResponse<T> | undefined;
```

xirzec commented on sdk/tables/data-tables/review/data-tables.api.md:
```diff
@@ -304,11 +304,43 @@ export interface SignedIdentifier {
+}
+
+// @public
+export interface TableBatchEntityResponse {
+    body?: any;
+    etag?: string;
+    rowKey?: string;
+    status: number;
+}
```
should this be of type `T` so you can type the response?

```suggestion
export interface TableBatchEntityResponse<T> {
    body?: T;
    etag?: string;
    rowKey?: string;
    status: number;
}
```

joheredi replied:
Thanks for bringing this up! the current implementation will never return content in the sub-requests unless it is an error. So I'm removing body throwing the error instead

If we decide to support query in the batch, it is possible that we'll want to support getting a body. At that time we should discuss how to type the bodies, as each sub-request may have different shapes

xirzec commented on sdk/tables/data-tables/src/TableBatch.ts:
```diff
@@ -0,0 +1,294 @@
+    const { span, updatedOptions } = createSpan("TableBatch-submitBatch", {});
+    const request = new WebResource(this.url, "POST", body, undefined, new HttpHeaders(headers));
+
+    request.spanOptions = updatedOptions;
```
`createSpan` returns `OperationOptions`, right? So shouldn't this be

```suggestion
    request.spanOptions = updatedOptions.spanOptions;
```

I'm kinda surprised that `{}` is allowed to satisfy `extends OperationOptions`, but I suppose it's because it's entirely optional.

joheredi replied:
oh! good catch thanks!

xirzec commented on sdk/tables/data-tables/src/TableBatch.ts:
```diff
@@ -0,0 +1,294 @@
+  const status = response.status;
+  const rawBody = response.bodyAsText || "";
+  const splitBody = rawBody.split(subResponsePrefix); // string after ending is useless
+  const subResponses = splitBody.slice(1, splitBody.length - 1); // string before first response boundary is useless
```
To make sure I understand the intention here, we're getting rid of the first and last item of the array, correct? Is that because the last item is an empty string after the final boundary?

joheredi replied:
Yes the first and last are the boundaries

joheredi replied:
I'm adding a comment

xirzec commented on sdk/tables/data-tables/src/TableBatch.ts:
```diff
@@ -0,0 +1,294 @@
+    const bodyMatch = response.match(/\{(.*)\}/);
+    const body = bodyMatch?.length === 2 ? JSON.parse(`${bodyMatch[0]}`) : undefined;
+
+    if (statusMatch?.length !== 2) {
+      throw new Error(`Couldn't extract status from sub-response:\n ${response}`);
+    }
+
+    const status = Number.parseInt(statusMatch[1]);
+    if (!Number.isInteger(status)) {
+      throw new Error(`Expected sub-response status to be an integer ${status}`);
+    }
```
should we do these checks on status immediately after matching it so we don't waste effort matching the other headers or calling JSON.parse on the body?

xirzec commented on sdk/tables/data-tables/src/TableBatchPolicies.ts:
```diff
@@ -0,0 +1,75 @@
+
+export class BatchHeaderFilterPolicyFactory implements RequestPolicyFactory {
+  // eslint-disable-next-line @azure/azure-sdk/ts-use-interface-parameters
+  public create(nextPolicy: RequestPolicy, options: RequestPolicyOptions): BatchHeaderFilterPolicy {
```
Personally I prefer to have factories return the base type of `RequestPolicyFactory` because then we don't have to expose the policy constructor directly if we ever want to expose the factory.

It's not a big deal since you don't actually export either right now, but might be helpful if these ever get moved elsewhere.

xirzec commented on sdk/tables/data-tables/src/TableBatchPolicies.ts:
```diff
@@ -0,0 +1,75 @@
+    nextPolicy: RequestPolicy,
+    // eslint-disable-next-line @azure/azure-sdk/ts-use-interface-parameters
+    options: RequestPolicyOptions
+  ): BatchRequestAssemblePolicy {
```
same remark here

xirzec reviewed (approved):
This is looking really good! 

I left a few last comments to be addressed, but great job on this! 👍

joheredi commented:
/azp run js - tables - tests

joheredi commented:
/azp run js - tables - tests
